File: 백준/Silver/15724. 주지수/주지수.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *

#setrecursionlimit(10**9)
input = stdin.readline

n,m = map(int,input().split())
area = [[0 for _ in range(m+1)]]

for _ in range(n):
    area.append([0]+list(map(int,input().split())))

dp = [[0]*(m+1) for _ in range(n+1)]
for i in range(1,n+1):
    for j in range(1,m+1):
        dp[i][j] = dp[i-1][j] + dp[i][j-1] - dp[i-1][j-1] + area[i][j] # 바로 위 + 바로 왼쪽 - 왼쪽위 대각선 + 현재 

for _ in range(int(input())):
    x,y,x2,y2 = map(int,input().split())
    
    x -=1
    y -=1

    print(dp[x2][y2] - dp[x][y2] - dp[x2][y] + dp[x][y]) 

File: 백준/Silver/15724. 주지수/주지수.swift
let nm = readLine()!.split{$0 == " "}.map{Int($0)!}
let (n,m) = (nm[0],nm[1])

var board: [[Int]] = [[Int](repeating: 0, count: m+1)]

for _ in 0..<n{
    board.append([0] + readLine()!.split{$0 == " "}.map{Int($0)!})
}

var dp: [[Int]] = [[Int]](repeating: [Int](repeating: 0, count: m+1), count: n+1)

for i in 1...n {
    for j in 1...m {
        dp[i][j] = dp[i-1][j] + dp[i][j-1] - dp[i-1][j-1] + board[i][j]
    }
}

let k = Int(readLine()!)!

for _ in 0..<k {
    let input = readLine()!.split{$0 == " "}.map{Int($0)!}
    
    let (x1,y1,x2,y2) = (input[0],input[1],input[2],input[3])
    
    
    print(dp[x2][y2] - dp[x2][y1-1] - dp[x1-1][y2] + dp[x1-1][y1-1])
    
}
